### ISSUE TITLE
Vision: Pallet Based SDK -> Network-based SDK 

### ISSUE THREAD

**Issue by kianenigma** [polkadot-api/polkadot-api#689]
> In the short term, treating it this way might work - as we are already doing with the pallet-based SDKs.

Is what I read in #666, and it gave me the idea to take this idea even further, and call it a network based SDK. 

I think a pallet-based SDK is mostly what can be implemented as runtime-apis. Where an SDK can then truly shine is to build on top of these runtime-apis, and provide an SDK over the entire polkadot network, not just a specific chain.  

For example, in the case of conviction voting, I am down to build a prototype that uses the `view-function` method introduced here, and does all the complicated logic of calculating convoluted voting mechanisms in the wasm blob. 

Full writeup: 

Let's see where this goes. Even if the network based SDK is too much, I am still convinced that to build good pallet-based APIs, we should use more code in the runtime as well.

**Comment by IkerAlus** [comment]
I believe that given the "multichain" capabilities of PAPI, it is the best suited (TS) library for this vision of a network-based SDK. However, given the discussion in https://github.com/polkadot-api/polkadot-api/issues/666 and others, I still wonder where we should put the line between on-chain logic (runtime APIs, view functions etc) and off-chain logic (higher level libraries, APIs etc). 

Let's consider the example @kianenigma in the blogpost above:

```js
dotApi = await new PolkadotNetwork(); 
{ transferrable, reserved } = dotApi.balance_of("kian");
```

Probably this presents the most basic functionality a user might be interested in, the one answering the question "how much DOT do I have in this entire network and how much of this can I transfer right now?"

But if you delve into what it actually implies, the point gets way more complex, for example:
- Do we assume that we have access to People chain storage to know the account owned by "kian"?
- Do the Network SDK know about all the potential SS58 prefixes? or are we assuming network-wide addresses?
- Are the concepts of transferrable, locked and reserved the same in all the chains that may potentially store DOT?
- Do all the relevant chains for this query have a light client implementation? if not, do we have stable enough RPC endpoints?
- and many more. 

With this I want to say that, whereas this Network-based SDK makes all the sense from the UX standpoint, it requires proper consensus, standarization and collaboration among core devs (Polkadot and parachains), API devs and application / front end devs. The question is: how can we achieve this liaison?

**Comment by joepetrowski** [comment]
> Do we assume that we have access to People chain storage to know the account owned by "kian"?

Yes, the network API would probably need some config file by the user to say which chains to include. And also the underlying chains would need some common APIs. At least in my early imagination of this there are three basic levels to the stack:

1. Pallet functions: APIs defined within pallets that never break so as not to expose storage formats.
2. Runtime functions: Bit of a collision with "runtime API", but I mean an API defined at the chain level. Something like [query account balances](https://github.com/polkadot-fellows/runtimes/blob/24bd45f/system-parachains/asset-hubs/asset-hub-polkadot/src/lib.rs#L1292), which cannot be defined at the pallet level, but rather aggregates information from several pallets into one response (in this case, all the assets an account owns on AH).
3. Network functions: What Kian is describing. It's built on top of the lower two, but the consumer might need to make some configs (maybe they don't want to query every chain for example).

> Do the Network SDK know about all the potential SS58 prefixes? or are we assuming network-wide addresses?

Kind of irrelevant. Prefixes are just a UI add on.

> Are the concepts of transferrable, locked and reserved the same in all the chains that may potentially store DOT?

Not necessarily, good point. But to be part of the network API, the chain would need to express something like "transferrable" and whatever else is defined.

> Do all the relevant chains for this query have a light client implementation? if not, do we have stable enough RPC endpoints?

This would just have to be part of the config of the tool.

**Comment by kianenigma** [comment]
> Do we assume that we have access to People chain storage to know the account owned by "kian"?


Yes, the network in this "Network-based API" is the Polkadot network, so it would mean the Polkadot RC + All system chains. 

> Yes, the network API would probably need some config file by the user to say which chains to include. And also the underlying chains would need some common APIs

@joepetrowski I see this as a more advance feature. Foremost, we should make it dead simple to build applications on top of DOT. In other words, the default for this config should be RC + all system chains. I like the extensibility idea though. 

Put differently, I would first put my effort into just the pink bubble: 

![Pasted image 20240904110357](https://github.com/user-attachments/assets/8c2e3515-13e5-4e6f-bce8-c6b9a75f3c9c)

The first and probably the only instance where we want to escape this bubble would be to give the user a sum of their DOT token across all parachains. 

> With this I want to say that, whereas this Network-based SDK makes all the sense from the UX standpoint, it requires proper consensus, standarization and collaboration among core devs (Polkadot and parachains), API devs and application / front end devs. The question is: how can we achieve this liaison?

Achieving this is mainly why I suggest we first implement this for the scope of the Polkadot Network, not "Polkadot Ecosystem" (which would include all parachains too). 

If we set a good example + have extensibility in mind, we would be able to then give clear instructions to other ecosystem teams as to how they should be able to tap into the same API ecosystem.
